ci-operator/step-registry/cucushift/upgrade/rollback/cucushift-upgrade-rollback-commands.sh

```diff
@@ -0,0 +1,139 @@
+    <failure message="">openshift cluster upgrade failed</failure>
+  </testcase>
+</testsuite>
+EOF
+    fi
+}
+
+function run_command() {
+    local CMD="$1"
+    echo "Running command: ${CMD}"
+    eval "${CMD}"
+}
+
+# Rollback the cluster to target release
+function rollback() {
+    res=$(run_command "oc adm upgrade rollback")
+    out="Requested rollback from ${TARGET_VERSION} to ${SOURCE_VERSION}"
+    if [[ ${res} == *"${out}"* ]]; then
+        echo "Rolling back cluster gets started..."
+    else
+        echo "Rolling back cluster doesn't start..."
+    fi
+}
+
+# Monitor the rollback status
+function check_rollback_status() {
+    local wait_rollback="${TIMEOUT}" out avail progress cluster_version
+    cluster_version="${TARGET_VERSION}"
+    echo "Starting the rollback checking on $(date "+%F %T")"
+    while (( wait_rollback > 0 )); do
+        sleep 5m
+        wait_rollback=$(( wait_rollback - 5 ))
+        if ! ( run_command "oc get clusterversion" ); then
+            continue
+        fi
+        if ! out="$(oc get clusterversion --no-headers || false)"; then
+            echo "Error occurred when getting clusterversion"
+            continue 
+        fi
+        avail="$(echo "${out}" | awk '{print $3}')"
+        progress="$(echo "${out}" | awk '{print $4}')"
+        if [[ ${avail} == "True" && ${progress} == "False" && ${out} == *"Cluster version is ${cluster_version}" ]]; then
+            echo -e "Rollback succeed on $(date "+%F %T")\n\n"
+            return 0
+        fi
+    done
+    if [[ ${wait_rollback} -le 0 ]]; then
+        echo -e "Rollback timeout on $(date "+%F %T"), exiting\n" && return 1
+    fi
+}
+
+# Check version, state in history
+function check_history() {
+    local version state
+    version=$(oc get clusterversion/version -o jsonpath='{.status.history[0].version}')
+    state=$(oc get clusterversion/version -o jsonpath='{.status.history[0].state}')
+    if [[ ${version} == "${TARGET_VERSION}" && ${state} == "Completed" ]]; then
+        echo "History check PASSED, cluster is now rollbacked to ${TARGET_VERSION}" && return 0
+    else
+        echo >&2 "History check FAILED, cluster rollbacked to ${TARGET_VERSION} failed, current version is ${version}, state is ${state}, exiting" && return 1
+    fi
+}
+
+if [[ -f "${SHARED_DIR}/kubeconfig" ]] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+fi
+
+# Setup proxy if it's present in the shared dir
+if [[ -f "${SHARED_DIR}/proxy-conf.sh" ]]; then
+    # shellcheck disable=SC1091
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+run_command "oc get machineconfig"
+
+export TARGET="${OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE}"
+TARGET_VERSION="$(env "NO_PROXY=*" "no_proxy=*" oc adm release info "${TARGET}" --output=json | jq -r '.metadata.version')"
+
+SOURCE_VERSION="$(oc get clusterversion --no-headers | awk '{print $2}')"
+SOURCE_MINOR_VERSION="$(echo "${SOURCE_VERSION}" | cut -f2 -d.)"
```

Comment:
It looks like `SOURCE_MINOR_VERSION` and `TARGET_MINOR_VERSION` are not used in any other places except in log, so could be dropped, right?

Reply:
Updated.